openscm/adapters/dice.py

```diff
@@ -0,0 +1,303 @@
+
+
+class DICE(Adapter):
+    """
+    Climate component from the Dynamic Integrated Climate-Economy (DICE) model by
+    William Nordhaus
+
+    https://sites.google.com/site/williamdnordhaus/dice-rice
+
+    TODO recalculate increase to absolute values
+    TODO set force[0]
+    TODO change to fluxes
+    TODO add array handling reading and writing timeseries
+    """
+
+    _M_atm0: float
+    """Initial concentration in atmosphere 2010 (GtC)"""
+
+    _M_atm_eq: float
+    """Equilibrium concentration atmosphere (GtC)"""
+
+    _M_atm_lower: float
+    """Lower bound concentration in atmosphere 2010 (GtC)"""
+
+    _M_l0: float
+    """Initial concentration in lower strata 2010 (GtC)"""
+
+    _M_l_eq: float
+    """Equilibrium concentration in lower strata (GtC)"""
+
+    _M_l_lower: float
+    """Lower bound for concentration in lower strata 2010 (GtC)"""
+
+    _M_u0: float
+    """Initial concentration in upper strata 2010 (GtC)"""
+
+    _M_u_eq: float
+    """Equilibrium concentration in upper strata (GtC)"""
+
+    _M_u_lower: float
+    """Lower bound for concentration in upper strata 2010 (GtC)"""
+
+    _T_atm0: float
+    """Initial atmospheric temp change (°C from 1900)"""
+
+    _T_atm_upper: float
+    """Upper bound for atmospheric temp change (°C from 1900)"""
+
+    _T_ocean0: float
+    """Initial lower stratum temp change (°C from 1900)"""
+
+    _T_ocean_lower: float
+    """Lower bound for lower stratum temp change (°C from 1900)"""
+
+    _T_ocean_upper: float
+    """Upper bound for lower stratum temp change (°C from 1900)"""
+
+    _b12: float
+    """Carbon cycle transition matrix"""
+
+    _b23: float
+    """Carbon cycle transition matrix"""
+
+    _c1: float
+    """Climate equation coefficient for upper level"""
+
+    _c3: float
+    """Transfer coefficient upper to lower stratum"""
+
+    _c4: float
+    """Transfer coefficient for lower level"""
+
+    _fex0: float
+    """2010 forcings of non-CO2 GHG (Wm-2)"""
+
+    _fex1: float
+    """2100 forcings of non-CO2 GHG (Wm-2)"""
+
+    _fco22x: float
+    """Forcings of equilibrium CO2 doubling (Wm-2)"""
+
+    _period_length: int
+    """Period length in seconds (currently 1yr)"""
+
+    _t2xco2: float
+    """Equilibrium temp impact (°C per doubling CO2)"""
+
+    _timestep: int
+    """Current time step"""
+
+    _timestep_count: int
+    """Total number of time steps"""
+
+    _year2100: int
+    """Timestamp of 2100-01-01"""
+
+    @abstractmethod
+    def initialize(self) -> None:
+        """
+        Initialize the model.
+        """
+
+        self._period_length = constants.YEAR
+        self._timestep = 0
+        self._timestep_count = (
+            self._stop_time - self._start_time
+        ) // self._period_length
+        self._year2100 = datetime(2100, 1, 1).timestamp()
+
+        # Read model parameters
+        model_parameter_defaults = {
+            # Initial concentration in atmosphere 2010 (GtC)
+            "M_atm0": (830.4, "GtC/a"),
```

Comment:
Is there a way to avoid writing the defaults in two places (as is currently effectively the case with this dictionary plus the list above)?

Reply:
Sorry, I requested the review too soon. Will do